api/src/main/java/jakarta/mail/Folder.java

```diff
@@ -651,6 +651,7 @@ public abstract boolean delete(boolean recurse)
     @Override
     public void close() throws MessagingException {
         close(true);
+        q.terminateQueue();
```

Comment:
[Folder.close(boolean)](https://github.com/eclipse-ee4j/angus-mail/blob/c959e0218111c39c986431299c1be2785f4feb6d/providers/imap/src/main/java/org/eclipse/angus/mail/imap/IMAPFolder.java#L1519) is not idempotent. I think where this is going is the right approach but it is plausible  to close(boolean) first followed by close() which now must be called but will fail.

Reply:
In the javadoc it says:
`This implementation calls <code>close(true)</code>.`
And as you pointed out, IMAPFolder will throw this if you invoke it 1+ times:
```
                throw new IllegalStateException(
                        "This operation is not allowed on a closed folder"
```
I guess this is a bug in IMAPFolder, that is not idempotent.

Reply:
The behavior is compliant with AutoCloseable interface but we are now forcing clients to call close twice  which doesn't work well with not idempotent. Just making the no arg close idempotent would be enough of a fix.

Reply:
Also should try close(true) and finally terminate.

Reply:
Done

Reply:
I think the tricky part of this is the user will do:

````
try(Folder f = ...) {
   //delete messages then...
   f.close(false);
} //<-- eventually calls close(true)
````

The test should be testing for this case. Also need to make sure an early close doesn't prevent termination of queue in 2nd call.  It looks like code fails to terminate because guard variable is scoped too wide.

Reply:
I am not sure how to handle this. Previously to my change, `Folder#close` was not idempotent either.

Now we are making it idempotent, but there are some combinations that we should allow.

My understanding is that we should only allow one time close(true). `q.terminateQueue();` is idempotent already. Then `#close`  would look like this:

```
    public void close() throws MessagingException {
            try {
                close(true);
            } finally {
                q.terminateQueue();
            }
    }
```
And the class that implements `#close(expugne)` should control that `#close(true)` is idempotent.

Reply:
Right, this is consistent with the javadoc:
`@throws IllegalStateException if this folder is not opened`

I have submitted it, thank you.

Reply:
You might want to add a test for 
````
try(Folder f = ...) {
   f.close(false);
} //<-- eventually calls close(true)
````

To ensure this completes normally.

Reply:
This depends on the implementation. We should test with IMAPFolder, but this is in Angus Mail and we don't have the dependency in in API. I can try to replicate IMAPFolder in the test.

Reply:
I missed that javadoc should be updated to remove.

````
@throws IllegalStateException if this folder is not opened
````

Reply:
The close with no arg still can throw IllegalStateException (it is catching it and rethrowing it if it is not opened). This is how it looks:

```
    @Override
    public void close() throws MessagingException {
        try {
            close(true);
        } catch (IllegalStateException ise) {
            if (isOpen()) {
                 throw ise;
            }
        } finally {
            q.terminateQueue();
        }
    }
```
Did I misunderstand something?.

Reply:
The patched behavior of no arg close is correct because it swallows ISE if folder is closed. Documentation is now wrong because it says it throws if closed "not open" but clearly we are throwing if open (which is correct).

We should drop the throws ISE from the docs because the catch code means "If we have caught an ISE and folder is open then the ISE is some is not API specified by close(boolean) so propagate it". No point in documenting ISE in no arg close because ISE is just like any propagated exception now.  If anything we could drop throws and and text about swallowing ISE. dropping should be enough.